### DISCUSSION TITLE
`COM812` with `black --preview`

### DISCUSSION THREAD

**Discussion by jamesbraza** [astral-sh/ruff#4213] [General]
With `black==23.3.0` and `ruff==0.0.264` and the below snippet:

```python
# 1: start
raise ValueError(
    "I'm a 1 very 2 very 3 very 4 very 5 very 6 very 7 very 8 very 9 very 10 very 11 very "
    "12 very long message"
)
```

Running `black --preview a.py`, no changes.

Next running `ruff --fixable=COM812 --select=COM812 --fix a.py`, it adds the trailing comma:

```python
# 2: ruff --fixable=COM812 --select=COM812 --fix a.py
raise ValueError(
    "I'm a 1 very 2 very 3 very 4 very 5 very 6 very 7 very 8 very 9 very 10 very 11 very "
    "12 very long message",
)
```

Now running `black --preview a.py` again, we see `black` adds an extra set of parenthesis, and wrapped the text at 88-chars with the space up front (per [this comment](https://github.com/psf/black/issues/182#issuecomment-399508289)):

```python
# 3: black --preview a.py
raise ValueError(
    (
        "I'm a 1 very 2 very 3 very 4 very 5 very 6 very 7 very 8 very 9 very 10 very"
        " 11 very 12 very long message"
    ),
)
```

The file has now stabilized with respect to both `black` and `ruff`.  The end result is that:
1. `black` spaced over 4 characters for an enclosing `()`
2. `black`'s `--experimental-string-processing` (part of `--preview`) wrapped the text over

At first thought, this seems like `COM812` and `black --preview` results in wasting of whitespace.

What do you think?  Is there anything we can do here?

I am thinking about disabling `COM812` so I can use `black --preview` and not have this churn.

**Comment by MichaReiser** [comment]
Thanks for bringing this up.

I'm not very familiar with the rule nor black, so forgive me for the basic question. 
Does `COM812` enforce trailing commas in situations where black isn't adding them automatically or what's your motivation for relying on both black and the rule for trailing comma enforcement?

From the black documentation
> You might have noticed that closing brackets are always dedented and that a trailing comma is always added. Such formatting produces smaller diffs; when you add or remove an element, it’s always just one line. Also, having the closing bracket dedented provides a clear delimiter between two distinct sections of the code that otherwise share the same indentation level (like the arguments list and the docstring in the example above).

**Reply by jamesbraza** [reply]
Yeah in the original snippet of the OP, `black --preview` wouldn't add a trailing comma, `flake8-commas` does though.

As posted by @calumy below, this will not be an issue in the future, since `black` reverted the code leading to the double indent.  Cheers!

**Comment by calumy** [comment]
I believe the changes that Black makes were reverted in https://github.com/psf/black/pull/3640 as the change was "too destructive based on feedback" to make it into the 2024 stable style.

**Reply by jamesbraza** [reply]
Oh nice, I would agree with their decision to revert it.  Looks like it was last week, so it hasn't shown up in `--preview` yet. 
 Thanks for sharing this!
